draft-ietf-mls-protocol.md

```diff
@@ -1290,70 +1290,150 @@ Given these inputs, the derivation of secrets for an epoch
+the PSK security expires before the next update (shorter PSK lifetime than
+update), does that constitute a weaker security level -->
+
+<!-- OPEN ISSUE Define "security level", and what it means to match the
+security level of the ciphersuite used in the group. -->
+
+Each PSK in MLS has a type that designates how it was provisioned.
+External PSKs are provided by the application logic. Other resumption PSKs
+are derived from the MLS key schedule, and used in cases where it is
+necessary to authenticate a member's participation in a prior group state.
+In particular, in addition to external PSK types, a PSK derived from within MLS
+may be used in the following cases:
+
+  - Re-Initialization: If, during the lifetime of the group, a change in the
+    fixed group parameters becomes necessary, e.g. if the ciphersuite used
+    by the group is deprecated or if the protocol version should be upgraded,
+    a PSK can be used to carry entropy from the old group forward into a new
+    group with the desired parameters.
+
+  - Recovery: If the group state of one or more members of the group deviates
+    from the rest, they can be re-added to the group such that participants who
+    were group members at the time of deviation and are still in the group can
+    verify that the added member was previously in the group.
+
+  - Branching: A PSK may be used to bootstrap a subset of current group
+    members into a new group. This applies if a subset of current group
+    members wish to branch based on the current group state.
+
+The injection of one or more PSKs into the key schedule is signaled in two
+ways: 1) as an extension to a Commit message that initiates a new epoch and 2)
+in the `GroupSecrets` object of a Welcome message sent to new members
+added in that epoch.
+
+~~~~~
+enum {
+  external(0),
+  reinit(1),
+  branch(2),
+  (255)
+} PSKType;
+
+struct {
+  PSKType psktype;
+  select (psktype) {
+
+    case external:
+      opaque psk_id<0..255>;
+
+    case reinit:
+      opaque psk_group_id<0..255>;
+      uint64 psk_epoch;
+
+    case branch:
+      opaque psk_group_id<0..255>;
+      uint64 psk_epoch;
+  }
+  opaque psk_nonce<0..255>;
+} PreSharedKeyID;
+
+struct {
+    PreSharedKeyID psks<0..2^16-1>;
+} PreSharedKeys;
+~~~~~
+
+On receiving a Commit with a `pre_shared_keys` extension or a GroupSecrets
+object with the psk field set, the receiving Client
+includes them in the key schedule in the order listed in the commit. For
+internal PSKs, the psk is defined as the `recovery_secret` of the group and
+epoch specified in the `PreSharedKeyID` object. Finally, the `psk_secret` to be
+included in the Key Schedule is derived as follows (where `n` is the number of
+PSKs):
+
+~~~~~
+struct {
+    PreSharedKeyID;
+    uint16 index;
+    uint16 count;
+} PSKLabel;
+
+psk_input_[i] = KDF.Extract(0, psk_[i])
+psk_secret_[i] = ExpandWithLabel(psk_input_[i], "derived psk", PSKLabel, KDF.Nh)
+psk_secret     = psk_secret_[i] || ... || psk_secret_[n]
```

Comment:
Maybe for the sake of efficiency, concat all the `psk_[i]` before they're put into Extract?

Reply:
I'm not sure what the assumptions would be under which we could prove that secure. Other than a random oracle that is. Since I don't think it's likely that people will inject more than one PSK at once very often, I'd rather leave this as it is.